src/Lecture4.hs

```diff
@@ -134,8 +144,34 @@ errors. We will simply return an optional result here.
 🕯 HINT: Use the 'readMaybe' function from the 'Text.Read' module.
 -}
 
+split :: Eq a => a -> [a] -> [[a]]
+split _ [] = []
+split separ list = part : split separ remainder
+  where
+    (part, remainder) =
+      list
+        & dropWhile (== separ)
+        & break (== separ)
+
+parseProductName :: String -> Maybe String
+parseProductName name =
+  if all isSpace name
+    then Nothing
+    else Just name
+
+parseCost :: String -> Maybe Int
+parseCost cs = do
+  c <- readMaybe cs :: Maybe Int
+  if c >= 0 then return c else Nothing
+
 parseRow :: String -> Maybe Row
-parseRow = error "TODO"
+parseRow rs = case split ',' rs of
+  [ps, ts, cs] -> do
+    p <- parseProductName ps
+    t <- readMaybe ts :: Maybe TradeType
+    c <- parseCost cs
+    return Row {rowProduct = p, rowTradeType = t, rowCost = c}
+  _ -> Nothing
```

Comment:
You have an excellent monadic implementation 🔝 

I haven't explained the following during lectures but there's a nice way to write this parsing syntax using `do`-notation, `Monad` instance for `Maybe` and the `MonadFail` desugaring trick 🧙🏻 

You can read more about this trick here:

* https://kowainik.github.io/posts/haskell-mini-patterns#monadfail-sugar

Your impleemntation is already great. And after applying this programming pattern, you can make the code more uniform without needed the `case-of` expression:

```suggestion
parseRow rs = do
    [ps, ts, cs] <- Just $ split ',' rs
    p <- parseProductName ps
    t <- readMaybe ts :: Maybe TradeType
    c <- parseCost cs
    return Row {rowProduct = p, rowTradeType = t, rowCost = c}
```

You can also construct the record slightly easier by using the `RecordWildCards` language extension:

```haskell
parseRow rs = do
    [ps, ts, cs] <- Just $ split ',' rs
    rowProduct <- parseProductName ps
    rowTradeType <- readMaybe ts
    rowCost <- parseCost cs
    pure Row{..}
```

You can read more about it in my blog post:

* https://kodimensional.dev/recordwildcards